WalletWasabi/WabiSabi/Client/AmountDecomposer.cs

```diff
@@ -1,130 +1,67 @@
-				.Where(x => x.Value > 1)
-				.Select(x => x.Key)
-				.ToArray();
+			RoundState = roundState;
 
-			List<Money> outputAmounts = new();
-			var remainingVsize = AvailableVsize;
+			// TODO limit to round imposed value
+			var maxEffectiveValue = possibleInputCoins.Sum(x => x.EffectiveValue(RoundState.FeeRate));
 
-			bool end = false;
-			foreach (var denom in denoms.Where(x => x <= remaining))
-			{
-				while (denom <= remaining)
-				{
-					if (remaining < MinimumAmountPlusFee || remainingVsize < 2 * OutputSize)
-					{
-						end = true;
-						break;
-					}
+			var allowedStandardValues = StandardDenomination.Values.Where(x => roundState.CoinjoinState.Parameters.AllowedOutputAmounts.Contains(x));
 
-					outputAmounts.Add(denom - OutputFee);
-					remaining -= denom;
-					remainingVsize -= OutputSize;
-				}
+			// TODO allow more than 6 outputs... but when?
+			var maxOutputs = Math.Min(maxAvailableVsize / Constants.P2WPKHOutputSizeInBytes, 6);
 
-				if (end)
-				{
-					break;
-				}
-			}
+			Logger.LogDebug($"Computing possible decompositions up to {maxOutputs} outputs ({maxAvailableVsize} vbytes available) for {maxEffectiveValue}");
 
-			if (remaining >= MinimumAmountPlusFee)
-			{
-				outputAmounts.Add(remaining - OutputFee);
-			}
+			PossibleDecompositions = new PossibleDecompositions(
+				allowedStandardValues,
+				maxEffectiveValue, // TODO needs to be higher to evaluate other users' inputs
+				Money.Zero,
+				maxOutputs);
 
-			return outputAmounts;
+			Logger.LogDebug($"At {FeeRate.Zero}, best decomposition is {string.Join(' ', PossibleDecompositions.GetByTotalValue().First().Outputs)}");
 		}
 
-		private Dictionary<Money, long> GetDenominationProbabilities(IEnumerable<Coin> allInputCoins)
-		{
-			var secondLargestInput = allInputCoins.OrderByDescending(x => x.Amount).Skip(1).FirstOrDefault();
-			IEnumerable<Money> demonsForBreakDown = StandardDenominationsPlusFee.Where(x => secondLargestInput is null || x <= secondLargestInput.EffectiveValue(FeeRate));
+		public RoundState RoundState { get; }
 
-			Dictionary<Money, long> denomProbabilities = new();
+		private PossibleDecompositions PossibleDecompositions { get; }
 
-			foreach (var input in allInputCoins)
-			{
-				foreach (var denom in BreakDown(input, demonsForBreakDown))
-				{
-					var weight = Weight(denom.Satoshi);
-
-					if (!denomProbabilities.TryAdd(denom, weight))
-					{
-						denomProbabilities[denom] += weight;
-					}
-				}
-			}
-
-			return denomProbabilities;
-		}
-
-		private long Weight(long val)
+		public IEnumerable<Money> Decompose(IEnumerable<Coin> myInputCoins, IEnumerable<Coin> allInputCoins, int availableVsize)
 		{
-			// Bias denom selection as the square of the value.
-			return val * val;
-		}
+			var totalEffectiveValue = myInputCoins.Select(x => x.EffectiveValue(RoundState.FeeRate)).Sum();
+			var minimumTotalValue = Money.Min(totalEffectiveValue.Percentage(99.99m), totalEffectiveValue - 5000L);
```

Comment:
This is worrisome. This means any user can lose around $3 randomly with today's BTC prices in every single mixing round. Worse, if someone mixes 1 BTC, then the random loss is $6 already. Expecting a 10x increase in BTC prices within a few years that's $60. 10 BTC $600, 100BTC $6000.